Add floatingIP allocation for LB creation on openstack

slaws opened the pull request:
This PR may relate to issue #12284 .
If a `floating-network-id` is added to the `[LoadBalancer]` section of openstack's cloud provider config, it will associate a new floatingIP to the VIP created.

Without a `floating-network-id` behaviour should remain the same.

slaws committed:
Add floatingIP allocation for LB creation on openstack

slaws commented:
I signed it!

cjcullen commented:
@slaws is your local git email address the same as your github email address?
(what does `git config --global user.email` say?)

If you want to keep those two addresses different, you will probably have to sign the CLA with your local git email as well.

Then again our CLA-bot has been known to complain for no reason...

cjcullen commented:
cc @anguslees who wrote this code originally.

slaws commented:
@cjcullen : it was the wrong mail indeed. 
I've added it as a secondary mail on my account, I don't know if this is enough.

slaws committed:
Fix : do not try to delete floatingIP if it does not exists

slaws commented:
@anguslees : Sorry I may have made a mistake by rebasing before correcting the point you spotted... Your comment has disappeared

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -639,9 +676,30 @@ func (lb *LoadBalancer) CreateTCPLoadBalancer(name, region string, externalIP ne
 	status := &api.LoadBalancerStatus{}
-	status.Ingress = []api.LoadBalancerIngress{{IP: vip.Address}}
+
+	if lb.opts.FloatingNetworkId == "" {
```
(style) I'd invert these, to avoid the branches diverging further over time:

```go
status.Ingress = []api.LoadBalancerIngress{{IP:vip.Address}}

if lb.opts.FloatingNetworkId != "" {
    // floating ip logic below
    // floatIP := ...
    status.Ingress = append(status.Ingress, {IP: floatIP.FloatingIP})
}

return status, nil

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -639,9 +676,30 @@ func (lb *LoadBalancer) CreateTCPLoadBalancer(name, region string, externalIP ne
+		return status, nil
+	}
+
+	vipAddr, err := getVipByName(lb.network, name)
```
We already have `vip` above, which I think should already include `vip.PortID`?

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -713,6 +771,19 @@ func (lb *LoadBalancer) EnsureTCPLoadBalancerDeleted(name, region string) error
+			return err
+		}
+		if floatingIP != nil {
+			error := floatingips.Delete(lb.network, floatingIP.ID).ExtractErr()
```
(minor) use `err` here to be consistent with everywhere else.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -639,9 +676,30 @@ func (lb *LoadBalancer) CreateTCPLoadBalancer(name, region string, externalIP ne
+	}
+	floatIPOpts := floatingips.CreateOpts{
+		FloatingNetworkID: lb.opts.FloatingNetworkId,
+		PortID:            vipAddr.PortID,
```
Aside: I suspect following https://github.com/kubernetes/kubernetes/pull/13005, that we want to use `externalIP` (now renamed to `LoadBalancerIP` elsewhere) to set this floating IP instead of the VIP (earlier in this function).  I think we should hold off on that however until we can see what the documentation leads users to expect - and what the other cloud providers end up implementing...

anguslees commented:
A few minor comments, otherwise lgtm

slaws committed:
Code review correction

Inverting code path on CreateTcploadBalancer to avoid branch divergence
Removing useless variable vipAddr as vip have information needed
Renaming 'error' variable on  EnsureTCPLoadBalancerDeleted to be consistent

slaws commented:
@anguslees Feedbacks should have been addressed in this commit.

slaws commented:
@anguslees @thockin could you please take another look on that PR and tell me if I missed something ?

anguslees commented:
lgtm

slaws commented:
@thockin is there anything I can do to submit this PR for merge ?

wojtek-t commented:
@k8s-bot ok to test

wojtek-t commented:
@k8s-bot unit test this please
